In [1]:
import sys
import sklearn
import numpy as np
import pandas as pd
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score
from sklearn.preprocessing import OneHotEncoder
from sklearn import preprocessing
from sklearn.model_selection import cross_val_score
import csv

# Data Cleaning

In [2]:
# Helper functions
def convert_si_to_number(x):
    total_stars = 0
    if 'k' in x:
        if len(x) > 1:
            total_stars = float(x.replace('k', '')) * 1000 # convert k to a thousand
    elif 'M' in x:
        if len(x) > 1:
            total_stars = float(x.replace('M', '')) * 1000000 # convert M to a million
    elif 'B' in x:
        total_stars = float(x.replace('B', '')) * 1000000000 # convert B to a Billion
    else:
        total_stars = int(x) # Less than 1000
    
    return int(total_stars)

In [3]:
# Read file
filename = "30000-40000.csv"
file = open("../data/30000-40000.csv")
reader = csv.reader(file)
i = 0
y = []
X = []
names = []
for row in reader:
    if len(row) < 6: # exclude data missing fields
        continue
    name = row[0]
    location = row[1]
    score = row[2]
    reviews = row[3]
    salaries = row[4]
    interviews = row[5]
    if len(location.split(",")) < 2: # exclude data missing states or cities
        continue
    
    names.append(name)
        
    # X features
    feature = []
    city = location.split(",")[0]
    state = location.split(",")[1]
    feature.append(city)
    feature.append(state)
    feature.append(convert_si_to_number(reviews))
    if salaries == '--':
        feature.append(0)
    else:
        feature.append(convert_si_to_number(salaries))
    if interviews == '--':
        feature.append(0)
    else:
        feature.append(convert_si_to_number(interviews))
    X.append(feature)

    # y labels
    y.append(score)
    
    i = i + 1

In [4]:
# Convert to pandas dataframe
X = np.array(X)
y = np.array(y)
X = pd.DataFrame({'City': X[:, 0], 'State': X[:, 1], 'Reviews': X[:, 2], 'Salaries': X[:, 3], 'Interviews': X[:, 4]})
# X = pd.DataFrame({'Reviews': X[:, 2], 'Salaries': X[:, 3], 'Interviews': X[:, 4]})

In [30]:
# Drop rows with NA fields
X = X.dropna()
X.shape
X

City Interviews Reviews Salaries            State
0                 Tokyo         10      46       21            Japan
1              Hamilton          9      46       39          Bermuda
2              San Jose         15      45       42               CA
3        Spokane Valley          3      45      107               WA
4               Ashburn         18      45       88               VA
5               Houston          9      45      135               TX
6                Austin         15      45       28               TX
7                Woburn         17      44      877               MA
8       Fort Lauderdale          5      44       55               FL
9               Horsham         27      44       40   United Kingdom
10              Toronto         14      44       33           Canada
11         Redwood City          7      43       45               CA
12             New York          6      43       15               NY
13                Paris          5      42       18           France
14              Houston         13      41       27               TX
15           Scottsdale          4      41       32               AZ
16     South Plainfield          6      41     1300               NJ
17               London          4      41       33   United Kingdom
18              Milford         31      41       35               CT
19             Stamford         10      41       69               CT
20             New York         21      41       14               NY
21               Nashua          8      41       22               NH
22               Tai Po          9      40       31        Hong Kong
23          Cooper City          7      40       23               FL
24           Alpharetta          4      40      213               GA
25               Irvine          6      40       35               CA
26                  Zug          3      40       37      Switzerland
27              Nogales         14      40       19               AZ
28            San Diego          8      40       12               CA
29             Portland          4      40       12               OR
...                 ...        ...     ...      ...              ...
12021          New York         29      20        9               NY
12022        San Marcos          0      20       17               TX
12023    Kennett Square          2      20       15               PA
12024        Vero Beach          2      20       12               FL
12025         San Diego          7      20       10               CA
12026           Lubbock          2      20        3               TX
12027         Arlington          3      20       20               VA
12028        Centennial          2      20       12               CO
12029          Bellevue          3      20        8               WA
12030            Murray          2      20        8               UT
12031          Berkeley          0      20       10               CA
12032            Tacoma          6      20       21               WA
12033       Upper Darby          7      20        8               PA
12034      Jacksonville          1      20       10               FL
12035            Austin          0      20        0               TX
12036         Arlington          1      20       11               WA
12037           Wichita          1      20        8               KS
12038            Irvine          1      20       12               CA
12039            Austin          1      20        4               TX
12040         Grapevine          1      20       11               TX
12041            Pelham          0      20        9               AL
12042           Phoenix          6      20       14               AZ
12043          Bellwood          6      20       13               PA
12044         Beachwood          0      20        7               OH
12045            Folsom          6      20       10               CA
12046         Arlington          3      20        8               VA
12047         Arl

In [31]:
# Convert to numbers
le = preprocessing.LabelEncoder()
X_2 = X.apply(le.fit_transform)
enc = preprocessing.OneHotEncoder()
enc.fit(X_2)
onehotlabels = enc.transform(X_2).toarray()
onehotlabels.shape

City  Interviews  Reviews  Salaries  State
0      2206           2      162       105     45
1       890         116      162       187     10
2      1945          21      158       195     12
3      2095          56      158        11    106
4        85          32      158       282    102
5       986         116      158        42     96
6       107          21      158       142     96
7      2445          29      156       281     53
8       748          82      156       227     24
9       983          51      156       192    101
10     2214          18      156       165     15
11     1814         100      150       203     12
12     1513          91      150        56     70
13     1665          82      149        81     27
14      986          15      148       139     96
15     1998          71      148       162      5
16     2077          91      148        38     67
17     1208          71      148       165    101
18     1376          58      148       173     14
19     2103           2      148       251     14
20     1513          41      148        47     70
21     1473         107      148       113     66
22     2165         116      145       155     33
23      486         100      145       116     24
24       33          71      145       108     28
25     1014          91      145       173     12
26     2489          56      145       180     94
27     1535          18      145        90      5
28     1941         107      145        26     12
29     1755          71      145        26     77
...     ...         ...      ...       ...    ...
12021  1513          55       78       285     70
12022  1949           0       78        72     96
12023  1070          36       78        56     78
12024  2280          36       78        26     24
12025  1941         100       78         2     12
12026  1231          36       78       150     96
12027    79          56       78        98    102
12028   372          36       78        26     13
12029   165          56       78       269    106
12030  1459          36       78       269     99
12031   183           0       78         2     12
12032  2163          91       78       105    106
12033  2250         100       78       269     78
12034  1032           1       78         2     24
12035   107           0       78         0     96
12036    79           1       78        14    106
12037  2403           1       78       269     46
12038  1014           1       78        26     12
12039   107           1       78       191     96
12040   847           1       78        14     96
12041  1688           0       78       285      3
12042  1711          91       78        47      5
12043   168          91       78        36     78
12044   146           0       78       252     75
12045   734          91       78         2     12
12046    79          56       78       269    102
12047    79           1       78       237    102
12048  1791           0       78       237     63
12049  1339          82       78        72     12
12050   812          36       78         0      5

[12051 rows x 5 columns]

In [8]:
# One hot encoding
enc = preprocessing.OneHotEncoder()
enc.fit(X_2)
X_3 = enc.transform(X_2).toarray()
X_3.shape

(12051, 3259)

In [19]:
sum(X_3[4])

5.0

In [287]:
# Convert to floats
X_3 = X_3.astype(np.float)
y = y.astype(np.float)

# Modeling

In [288]:
# Test model using 5 fold cross validation
model = LinearRegression()
scores = cross_val_score(model, X_3, y, cv=5, scoring = "neg_mean_squared_error") # also use R^2
scores

array([-7.18030106e+24, -6.73594703e+24, -1.13052612e+25, -2.21435168e+24,
       -3.46957127e+24])

In [289]:
# Train model on training data
model = LinearRegression()
model.fit(X_3, y)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=1, normalize=False)

In [290]:
# Make predictions
predictions = model.predict(X_3)

In [291]:
# Calculate statistics on training data
error = mean_squared_error(y, predictions)
w = model.coef_
b = model.intercept_
error, w, b

(0.3102284362023129,
 array([ 8.50261952e+11, -5.50221018e+11, -5.50221018e+11, ...,
         1.75720673e+12,  1.75720673e+12,  1.75720673e+12]),
 -12612990674079.588)

# Results

Non-location features: 
error, w, b is (0.45079620978786505,
 array([ 0.00647003, -0.00115187, -0.00028681]),
 3.3909317414654145)

5-fold cross validation returns
array([-0.44058222, -0.46424376, -0.44806606, -0.44477971, -0.45780714])

 
All features: 
error, w, b is (0.3102284362023129,
 array([ 8.50261952e+11, -5.50221018e+11, -5.50221018e+11, ...,
         1.75720673e+12,  1.75720673e+12,  1.75720673e+12]),
 -12612990674079.588)

5-fold cross validation returns
array([-7.18030106e+24, -6.73594703e+24, -1.13052612e+25, -2.21435168e+24,
       -3.46957127e+24])

# Experimental (ignore)

In [99]:
import pandas as pd
X = pd.read_csv("../data/30000-40000.csv")


(12053, 6)

In [100]:
le = preprocessing.LabelEncoder()
X_2 = X.apply(le.fit_transform)
enc = preprocessing.OneHotEncoder()
enc.fit(X_2)
onehotlabels = enc.transform(X_2).toarray()
onehotlabels.shape

(12053, 15297)

In [124]:
data = np.array([[5.8, 2.8], [6.0, 2.2]])
dataset = pd.DataFrame({'Column1': data[:, 0], 'Column2': data[:, 1]})
dataset

Column1  Column2
0      5.8      2.8
1      6.0      2.2